### PR TITLE
[APINotes] Honor Swift 4 API notes in Swift 3 mode.

### PR DESCRIPTION
**Pull request by jrose-apple**
More generally, change the meaning of the SwiftVersions section to be "this version or earlier" rather than "exactly this version". This keeps us from having to duplicate information for builds of Swift that support more than two possible versions (e.g. 3, 4, and 5).

### CODE REVIEW

File: lib/APINotes/APINotesReader.cpp [hunk 1]
```diff
@@ -1492,8 +1492,13 @@ APINotesReader::VersionedInfo<T>::VersionedInfo(
...
-    if (Results[i].first == version) {
-      Selected = i;
+    if (version && Results[i].first >= version) {
+      // Pick the closest match. If the current version is "4", then entries for
+      // 4 are better than entries for 5, but both are valid.
+      if (Selected == Results.size() ||
+          Results[Selected].first > Results[i].first) {
+        Selected = i;
+      }
       break;
```
Comment (jrose-apple) [inline lib/APINotes/APINotesReader.cpp]: Oops. This `break` is wrong. Either I need a test to catch this, or I should guarantee that the list is always sorted.

File: lib/APINotes/APINotesWriter.cpp [hunk 2]
```diff
@@ -475,10 +475,16 @@ namespace {
   template<typename T>
   void emitVersionedInfo(
          raw_ostream &out,
-         const SmallVectorImpl<std::pair<VersionTuple, T>> &infoArray,
+         SmallVectorImpl<std::pair<VersionTuple, T>> &infoArray,
          llvm::function_ref<void(raw_ostream &out,
                                  const typename MakeDependent<T>::Type& info)>
            emitInfo) {
+    std::sort(infoArray.begin(), infoArray.end(),
```
Comment (DougGregor) [inline lib/APINotes/APINotesWriter.cpp:482]: No `array_pod_sort`, eh?

**Review by DougGregor — APPROVED** [review]
LGTM, nice testing.

File: lib/APINotes/APINotesWriter.cpp (hunk 2 above)
Comment (jrose-apple) [reply lib/APINotes/APINotesWriter.cpp:482]: It's not necessarily POD, is it?

Comment (DougGregor) [reply lib/APINotes/APINotesWriter.cpp:482]: Well. Maybe not by the language definition., but it'd work. `std::sort()` is not exactly a small thing to instantiate.

**Pull request merged**